### PR TITLE
Correct example in "Function arguments (2 or fewer ideally)"

### LINKED ISSUE #151: Disagree about: `Function arguments (2 or fewer ideally)`
I don't really believe that this is always a best practice.

IMHO the most of the cases are better to be explicit about all the dependencies needed by an object to be created, avoiding to created object in an invalid state.

Creating only valid object will reduce the amount of `if ($object->getProperty() !== null)` and will help you to better focus on the domain.

If the creation of an object is "hard" you can use a factory/builder to make it easier to interact with, but it will ALWAYS create an object in a valid state.

### PR DESCRIPTION
**Pull request by peter-gribanov**
fix #151

@damianopetrungaro what do you think about this change? I think that in this form using the DTO is more obvious.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -443,33 +443,34 @@ of the time a higher-level object will suffice as an argument.
...
 
 **Good:**
 
 ```php
 class MenuConfig
 {
-    public $title;
-    public $body;
-    public $buttonText;
+    public $title = '';
```
Comment (damianopetrungaro) [inline README.md]: THose by default are null, I don't see any benefit of setting them as a string :)

Comment (damianopetrungaro) [inline README.md:459]: THose by default are null, I don't see any benefit of setting them as a string :)

File: README.md [hunk 2]
```diff
@@ -443,33 +443,34 @@ of the time a higher-level object will suffice as an argument.
...
-$config->body = 'Bar';
-$config->buttonText = 'Baz';
-$config->cancellable = true;
-
 function createMenu(MenuConfig $config): void
 {
     // ...
 }
+
+$config = new MenuConfig();
```
Comment (damianopetrungaro) [inline README.md:470]: We should ensure that MenuCOnfig is in a valid state too.
So sth like

```

class DTO
{
    /**
     * @var string
     */
    private $title;

    /**
     * @var string
     */
    private $body;

    /**
     * @var string
     */
    private $buttonText;

    /**
     * @var bool
     */
    private $cancellable;

    public function __construct(string $title, string $body, string $buttonText, string $cancellable)
    {
        $this->title = $title;
        $this->body = $body;
        $this->buttonText = $buttonText;
        $this->cancellable = $cancellable;
    }
}
```

File: README.md (hunk 1 above)
Comment (peter-gribanov) [reply README.md:459]: The whole point of this change is that the initial value is not null.

File: README.md (hunk 2 above)
Comment (peter-gribanov) [reply README.md:470]: Not. It is not required to be valid. Do not confuse DTO and VO.

**Comment by damianopetrungaro** [comment]
There is a difference between VOs and simple objects in a valid state, the
first one  validate using the domain rules, the second one don't.

Valid state is not only related to domain object, it's a good practice in
general :)

**Comment by peter-gribanov** [comment]
@damianopetrungaro not. A valid state is useful for DTO, but not required.
Data Transfer Object (DTO) is only a means of organizing data, such as array. But unlike the array, the DTO has an explicit structure.
If we use array instead of DTO, is the array required to be valid? Yes, it is desirable, but it cannot be guaranteed and therefore we validate its structure and values ​​in the receiving side. With the DTO is the same, only we do not check the structure.

In your example, you made all the properties private and now you need to create getters and setters to access them, which turns a simple DTO into an Anemic Model.

```php
class DTO
{
    public $title = '';

    public $body = '';

    public $buttonText = '';

    public $cancellable = false;
}
```

```php
class DTO
{
    private $title;

    private $body;

    private $buttonText;

    private $cancellable;

    public function __construct(string $title = '', string $body = '', string $buttonText = '', string $cancellable = false)
    {
        $this->title = $title;
        $this->body = $body;
        $this->buttonText = $buttonText;
        $this->cancellable = $cancellable;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function setTitle(string $title): void
    {
        $this->title = $title;
    }

    public function getBody(): string
    {
        return $this->body;
    }

    public function setBody(string $body): void
    {
        $this->body = $body;
    }

    public function getButtonText(): string
    {
        return $this->buttonText;
    }

    public function setButtonText(string $buttonText): void
    {
        $this->buttonText = $buttonText;
    }

    public function getCancellable(): string
    {
        return $this->cancellable;
    }

    public function setCancellable(string $cancellable): void
    {
        $this->cancellable = $cancellable;
    }
}
```

**Comment by damianopetrungaro** [comment]
The DTO Is not a model, so the anemic anti pattern is not gonna play here :)

The setters are not required because you are injecting all the values at
the beginning during the object creation.

An array can't be in a valid state because you can't assure the valid state
(anyone everywhere can change the state of the array).

On the other hand an object with private fields and no setters can't be
changed :)

**Comment by peter-gribanov** [comment]
That is the meaning of the DTO that its state can be changed as well as the array.
The state of VO cannot be changed, but the state of the DTO can.

**Comment by damianopetrungaro** [comment]
DTO are short live object.
If you make it immutable you just ensure more security on your code.

Immutable > mutable ;)

**Comment by peter-gribanov** [comment]
I repeat again. DTO can be declared as immutable, but it is **not obliged** to be such. DTO can be mutable  and this is normal.
In [this case](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally), declaring the DTO as immutable makes its use meaningless.
I can make some example using VO, but this example will be about DDD, and not about Clean Code.

**Comment by damianopetrungaro** [comment]
Your repo, your role.
You keep mixing DDD concepts with clean code :)

But if you want code easy to change and they avoid possible side effects
that's the way.

Thank for answering quickly, nice conversation :)

**Comment by damianopetrungaro** [comment]
And again  (just to clarify), DTO should be immutable with only getters.
They are not "meaningless" and they have a precise and concise scope object
to pass data.
If you make them mutable then you won't know about the data anymore.

In your example I can simply switch any property to an integer and cause
error later :)

**Comment by peter-gribanov** [comment]
I'm talking about meaninglessness in a [specific example](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally).
If in [this example](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally) we make the DTO immutable, then in fact we will replace this:
```php
function createMenu(string $title, string $body, string $buttonText, bool $cancellable): void
```
to this:
```php
public function __construct(string $title, string $body, string $buttonText, string $cancellable)
```
and we return to the problem with which we started.

**Comment by damianopetrungaro** [comment]
This is because having too many argument is not a bad idea.
If you have a lot of dependencies you can't split them, handle them.

**Comment by peter-gribanov** [comment]
I propose to turn the original source.

> The ideal number of arguments for a function is zero (niladic). Next comes one (monadic), followed
closely by two (dyadic). Three arguments (triadic) should be avoided where possible. More than three
(polyadic) requires very special justification — and then shouldn’t be used anyway.
> *Clean Code book, Chapter 3, page 40*

I assume that [Function arguments (2 or fewer ideally)](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally) section is based on **Argument Objects** section from Clean Code book, Chapter 3, page 43.

> When a function seems to need more than two or three arguments, it is likely that some of
those arguments ought to be wrapped into a class of their own. Consider, for example, the difference between the two following declarations:
>
>```java
> Circle makeCircle(double x, double y, double radius);
> Circle makeCircle(Point center, double radius);
> ```
>
> Reducing the number of arguments by creating objects out of them may seem like
cheating, but it’s not. When groups of variables are passed together, the way `x` and
`y` are in the example above, they are likely part of a concept that deserves a name of its
own.

**Comment by peter-gribanov** [comment]
I create alternative solution with VO #168

**Comment by TomasVotruba** [comment]
Resolved in PR #168

**Pull request closed without merging**